File: 78_Subsets.java
class Solution {

    public List<List<Integer>> subsets(int[] nums) {
        ArrayList<List<Integer>> result = new ArrayList<>();
        ArrayList<Integer> current = new ArrayList<>();
        dfs(result, current, nums, 0);
        return result;
    }

    private void dfs(List<List<Integer>> result, List<Integer> current, final int[] nums, int index) {

        /* Termination */
        if(index == nums.length) {
            ArrayList<Integer> result_part = new ArrayList<>(current);
            result.add(result_part);
            return;
        }

        dfs(result, current, nums, index + 1);
        current.add(Integer.valueOf(nums[index]));
        dfs(result, current, nums, index + 1);
        current.remove(current.size() - 1);     /* List current have to be unchanged */

    }

}

File: 78_Subsets.cpp
class Solution {
  void dfs(vector<int>& current, vector<vector<int>>& result,
           const vector<int>& nums, int posStartFrom) {
    result.push_back(current);
    for (int i = posStartFrom; i < nums.size(); ++i) {
      current.push_back(nums[i]);
      dfs(current, result, nums, i + 1);
      current.pop_back();
    }
  }

 public:
  vector<vector<int>> subsets(vector<int>& nums) {
    vector<vector<int>> result;
    vector<int> current;
    sort(nums.begin(), nums.end());
    dfs(current, result, nums, 0);
    return result;
  }
};